File: core/contractsdk/go/example/hash_deposit/hash_deposit.go
package main

import (
	"github.com/xuperchain/xuperchain/core/contractsdk/go/code"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/driver"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/utils"
)

const (
	UserBucket = "USER"
	HashBucket = "HASH"
)

type hashDeposit struct {
}

func (hd *hashDeposit) Initialize(ctx code.Context) code.Response {
	return code.OK(nil)
}

func (hd *hashDeposit) StoreFileInfo(ctx code.Context) code.Response {
	args := struct {
		UsedID   []byte `json:"user_id",required:"true"`
		HashID   []byte `json:"hash_id",required:"true"`
		FileName []byte `json:"file_name",required:"true"`
	}{}
	err := utils.Validate(ctx.Args(), &args)
	if err != nil {
		return code.Error(err)
	}
	userKey := utils.Concat(UserBucket, "/", args.UsedID, "/", args.HashID)
	hashKey := utils.Concat(HashBucket, "/", args.HashID)
	value := utils.Concat(args.UsedID, "\t", args.HashID, "\t", args.FileName)

	//TODO
	// error == nil means hash exists already
	if _, err = ctx.GetObject(utils.Concat(hashKey)); err == nil {
		return code.Error(err)
	}
	if err := ctx.PutObject(userKey, value); err != nil {
		return code.Error(err)
	}

	if err := ctx.PutObject(hashKey, value); err != nil {
		return code.Error(err)
	}
	return code.OK(nil)
}

func (hd *hashDeposit) queryUserList(ctx code.Context) code.Response {
	key := utils.Concat(UserBucket, "/")
	iter := ctx.NewIterator(key, utils.Concat(key, "~"))
	result := []byte{}
	for iter.Next() {
		//TODO error 处理@fengjin
		k := string(iter.Key())
		v := iter.Value()
		if len(k) > len(UserBucket)+1 {
			result = append(result, utils.Concat(v[len(UserBucket)+1:], "\n")...)
		}
	}
	return code.OK(utils.Concat(result))
}

func (hd *hashDeposit) QueryFileInfoByUser(ctx code.Context) code.Response {
	args := struct {
		UserID []byte `json:"user_id",required:"true"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	result := []byte{}
	start := utils.Concat(UserBucket, "/", args.UserID)
	end := utils.Concat(start, "~")
	iter := ctx.NewIterator(start, end)
	for iter.Next() {
		//TODO error 处理 @fengjin
		result = append(result, utils.Concat(iter.Value(), "\n")...)
	}
	return code.OK(result)
}

func (hd *hashDeposit) QueryFileInfoByHash(ctx code.Context) code.Response {
	args := struct {
		HashID []byte `json:"hash_id",required:"true"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	key := utils.Concat(HashBucket, "/", args.HashID)
	value, err := ctx.GetObject(key)
	if err != nil {
		return code.Error(err)
	}
	return code.OK(value)
}

func main() {
	driver.Serve(new(hashDeposit))
}

File: core/contractsdk/java/example/hash_deposit/src/main/java/com/baidu/xuper/HashDeposit.java
package com.baidu.xuper;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Iterator;
import java.util.function.Consumer;

import com.baidu.xuper.*;
import com.baidu.xuper.ContractIteratorItem;

/**
 * Counter
 */
public class HashDeposit implements Contract {
        final byte [] Bucket = "USER".getBytes();
        final byte []HashBucket = "HASH".getBytes();

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    public Response storeFileInfo(Context ctx){
        byte[] user_id = ctx.args().get("user_id");
        byte[] hash_id = ctx.args().get("hash_id");
        byte[] file_name = ctx.args().get("file_name");
        byte[] userKey = null;
        byte[] hashKey = null;
        byte []value = null;
        if (ctx.getObject(hashKey)){
            return  Response.error("err");
        }
//        这里就不会出错吗 ？
        ctx.putObject(userKey,value);
        ctx.putObject(hashKey,value);
        return Response.ok("".getBytes());
    }

    @ContractMethod
    public Response queryUserList(Context ctx){
        String key = "";
        String start = "";
        String end = "";
       Iterator<ContractIteratorItem> iter = ctx.newIterator(start.getBytes(),end.getBytes());
       final ByteBuffer result = new ByteBuffer(0,0,0,0);
       iter.forEachRemaining(
               contractIteratorItem -> {
                   byte[]key = contractIteratorItem.getKey();
                   byte[]value = contractIteratorItem.getValue();
//                   if key.length > Userkey.length +1{
                   result.put(value);
//                   }
               }
       );
       return Response.ok(result.array());
    };

    @ContractMethod
    public Response queryFileInfoByUser(Context ctx){
        String key = "";
        String start = "";
        String end  = "";
//        并发安全性？？
        final ByteBuffer result = new ByteBuffer(0,0,0,0);
        ctx.newIterator(start.getBytes(),end.getBytes()).forEachRemaining(
                contractIteratorItem -> {
                    result.put(contractIteratorItem.getValue());
                    result.put("\n".getBytes());
                }
        );
        return Response.ok(result.array());

    }

    @ContractMethod
    public Response queryFileInfoByHash(Context ctx ){
        String key="";
        return Response.ok(ctx.getObject(key.getBytes()));
    }

    public static void main(String[] args) {
        Driver.serve(new HashDeposit());
    }
}